src/turbine/turbine_tree.zig

```diff
@@ -0,0 +1,910 @@
+        index = if (index == 0) index else index + offset;
+        return nodes[index].pubkey();
+    }
+
+    /// All staked nodes + other known tvu-peers + the node itself;
+    /// sorted by (stake, pubkey) in descending order.
+    fn getNodes(
+        allocator: std.mem.Allocator,
+        my_contact_info: ThreadSafeContactInfo,
+        tvu_peers: []const ThreadSafeContactInfo,
+        staked_nodes: *const std.AutoArrayHashMapUnmanaged(Pubkey, u64),
+        use_stake_hack_for_testing: bool,
+    ) !std.ArrayList(Node) {
+        var nodes = try std.ArrayList(Node).initCapacity(allocator, tvu_peers.len + staked_nodes.count());
+        defer nodes.deinit();
+
+        var pubkeys = std.AutoArrayHashMap(Pubkey, void).init(allocator);
+        defer pubkeys.deinit();
+
+        // Add ourself to the list of nodes
+        if (use_stake_hack_for_testing) {
+            var max_stake: u64 = 0;
+            for (staked_nodes.values()) |stake| if (stake > max_stake) {
+                max_stake = stake;
+            };
+            nodes.appendAssumeCapacity(.{ .id = .{ .contact_info = my_contact_info }, .stake = @divFloor(max_stake, 2) });
+        } else {
+            try nodes.append(.{
+                .id = .{ .contact_info = my_contact_info },
+                .stake = if (staked_nodes.get(my_contact_info.pubkey)) |stake| stake else 0,
+            });
+        }
+        try pubkeys.put(my_contact_info.pubkey, void{});
+
+        // Add all TVU peers directly to the list of nodes
+        // The TVU peers are all nodes in gossip table with the same shred version
+        for (tvu_peers) |peer| {
+            nodes.appendAssumeCapacity(.{
+                .id = .{ .contact_info = peer },
+                .stake = if (staked_nodes.get(peer.pubkey)) |stake| stake else 0,
+            });
+            try pubkeys.put(peer.pubkey, void{});
+        }
+
+        // Add all staked nodes to the list of nodes
+        // Skip nodes that are already in the list, i.e. nodes with contact info
+        for (staked_nodes.keys(), staked_nodes.values()) |pubkey, stake| {
+            if (stake > 0 and !pubkeys.contains(pubkey)) {
+                nodes.appendAssumeCapacity(.{
+                    .id = .{ .pubkey = pubkey },
+                    .stake = stake,
+                });
+            }
+        }
+
+        // Sort the nodes by stake, then pubkey
+        std.mem.sortUnstable(Node, nodes.items, {}, struct {
+            pub fn lt(_: void, lhs: Node, rhs: Node) bool {
+                if (lhs.stake > rhs.stake) return true;
+                if (lhs.stake < rhs.stake) return false;
+                return std.mem.lessThan(u8, &lhs.pubkey().data, &rhs.pubkey().data);
+            }
+        }.lt);
+
+        // Filter out nodes which exceed the maximum number of nodes per IP and
+        // nodes with a stake of 0
+        var result = try std.ArrayList(Node).initCapacity(allocator, nodes.items.len);
```

Comment:
maybe a better name for this -- `filtered/result_nodes` ? not sure

Reply:
2ee5728a